icu4c/source/test/intltest/collationtest.cpp

```diff
@@ -1925,4 +1939,187 @@ void CollationTest::TestBuilderContextsOverflow() {
+#endif
+
+    assertTrue("UnicodeString", equal_to(UnicodeString::readOnlyAlias(TEXT_CHAR16), TEXT_CHAR16));
+    assertTrue("string", equal_to(std::string{TEXT_CHAR}, TEXT_CHAR));
+}
+
+// Verify that the UCollator predicates handle all string types.
+void CollationTest::TestUCollatorPredicateTypes() {
+    using namespace U_HEADER_NESTED_NAMESPACE;
+    IcuTestErrorCode status(*this, "TestUCollatorPredicateTypes");
+    setRootCollator(status);
+    status.assertSuccess();
+    const auto equal_to = collator::equal_to(coll->toUCollator());
+
+    assertTrue("char16_t", equal_to(TEXT_CHAR16, TEXT_CHAR16));
+    assertTrue("u16string_view", equal_to(std::u16string_view{TEXT_CHAR16}, TEXT_CHAR16));
+
+#if !U_CHAR16_IS_TYPEDEF && (!defined(_LIBCPP_VERSION) || _LIBCPP_VERSION < 180000)
+    assertTrue("uint16_t", equal_to(TEXT_UINT16, TEXT_UINT16));
+    assertTrue("basic_string_view<uint16_t>",
+               equal_to(std::basic_string_view<uint16_t>{TEXT_UINT16}, TEXT_UINT16));
+#endif
+
+#if U_SIZEOF_WCHAR_T==2
+    assertTrue("wchar_t", equal_to(TEXT_WCHAR, TEXT_WCHAR));
+    assertTrue("wstring_view", equal_to(std::wstring_view{TEXT_WCHAR}, TEXT_WCHAR));
+#endif
+
+    assertTrue("char", equal_to(TEXT_CHAR, TEXT_CHAR));
+    assertTrue("string_view", equal_to(std::string_view{TEXT_CHAR}, TEXT_CHAR));
+
+#if defined(__cpp_char8_t)
+    assertTrue("char8_t", equal_to(TEXT_CHAR8, TEXT_CHAR8));
+    assertTrue("u8string_view", equal_to(std::u8string_view{TEXT_CHAR8}, TEXT_CHAR8));
+#endif
+
+    assertTrue("UnicodeString", equal_to(UnicodeString::readOnlyAlias(TEXT_CHAR16), TEXT_CHAR16));
+    assertTrue("string", equal_to(std::string{TEXT_CHAR}, TEXT_CHAR));
+}
+
+void CollationTest::TestCollatorMap() {
+    using namespace U_HEADER_NESTED_NAMESPACE;
+    IcuTestErrorCode status(*this, "TestCollatorMap");
+    setRootCollator(status);
+    status.assertSuccess();
+    coll->setStrength(Collator::PRIMARY);
+    std::map<std::string, int, decltype(coll->less())> m{coll->less()};
+    ++m["a"];
+    ++m["b"];
+    ++m["A"];
+    assertEquals("m.size()", 2, m.size());
+    assertEquals(R"(m["a"])", 2, m["a"]);
+
+    std::map<std::u16string, int, collator::less> u16m{collator::less(coll->toUCollator())};
+    ++u16m[u"a"];
+    ++u16m[u"b"];
+    ++u16m[u"A"];
+    assertEquals("u16m.size()", 2, u16m.size());
+    assertEquals(R"(u16m["a"])", 2, u16m[u"a"]);
+
+#if defined(__cpp_char8_t)
+    std::map<std::u8string, int, collator::less> u8m{collator::less(coll->toUCollator())};
+    ++u8m[u8"a"];
+    ++u8m[u8"b"];
+    ++u8m[u8"A"];
+    assertEquals("u8m.size()", 2, u8m.size());
+    assertEquals(R"(u8m["a"])", 2, u8m[u8"a"]);
+#endif
+
+    std::map<UnicodeString, int, collator::less> um{collator::less(coll->toUCollator())};
+    // Only UnicodeString allows heterogeneous lookup across encodings:
+    ++um[u"a"];
+    ++um["b"];
+    // ++um[u8"A"];  // TODO(egg): Should this be legal?
```

Comment:
`u8string_view` is C++20 though.

Reply:
That's fine. It could be conditional on that being available, like the char8_t test code that Fredrik is adding here.
Just a ticket for now, no more additions for 76 anyway.

Reply:
This is entirely a question about the `UnicodeString` constructor (which currently doesn't accept `char8_t*`), it has nothing to do with collation.

Reply:
I know that this is a side issue here. I asked Robin whether he could submit a ticket. Not seeing a response for that, I just created https://unicode-org.atlassian.net/browse/ICU-22895.